### ISSUE TITLE
cmd/compile: not inlined closures can't inline inner closures

### ISSUE THREAD

**Issue by DmitriyMV** [golang/go#41988]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.15 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes.

### What did you do?

Wrote code:

```go
package go_closures

func callInt(val int) int {
	getInt := func() int { return val }
	return callFn(getInt)
}

func callFn(getInt func() int) int {
	outer := func() int { return getInt() }
	return globalFunc(outer)
}

var globalFunc = func(f func() int) int {
	return f()
}
```
</details>

### What did you expect to see?

`getInt` is inlined into `outer`. `outer` escapes to the heap.

### What did you see instead?

`getInt` isn't inlined into `outer`. Both `outer` and `getInt` escape to the heap.

**Comment by DmitriyMV** [comment]
The main problem here is that `getInt` closure escapes when it can be folded into `outer` (even with call).

**Comment by mdempsky** [comment]
The issue here is we don't currently support inlining functions that contain closures. We'd first need to inline `callFn` into `callInt` to be able to inline `getInt` into `outer`.

Escape analysis is a red herring here. Escape analysis runs after inlining and doesn't affect inlining decisions.

**Comment by danscales** [comment]
Yes, I will look at this bug, after I submit the base change that allows inlining of functions with closures.

**Comment by DmitriyMV** [comment]
Still reproduces on 1.17:
```
package main

import (
	"fmt"
	"testing"
)

func callInt(val int) int {
	getInt := func() int { return val }
	return callFn(getInt)
}

func callFn(getInt func() int) int {
	outer := func() int { return getInt() }
	return globalFunc(outer)
}

var globalFunc = func(f func() int) int {
	return f()
}

func main() {
	result := testing.Benchmark(Bench)
	fmt.Println("result.AllocsPerOp()", result.AllocsPerOp())
}

func Bench(b *testing.B) {
	for i := 0; i < b.N; i++ {
		callInt(i)
	}
}

```

**Comment by danscales** [comment]
Thank you, will take a look if any improvement is possible for 1.18.

**Comment by DmitriyMV** [comment]
Looks like it's fixed on Go 1.23 - there is now only two allocations for variable and for function literal.
